Mage.Sets/src/mage/cards/r/RockHydra.java

```diff
@@ -0,0 +1,131 @@
+import mage.abilities.effects.common.PreventDamageToSourceEffect;
+import mage.abilities.effects.common.counter.AddCountersSourceEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.PhaseStep;
+import mage.constants.Zone;
+import mage.counters.CounterType;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.permanent.Permanent;
+
+/**
+ *
+ * @author anonymous
+ */
+public class RockHydra extends CardImpl {
+
+    public RockHydra(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{X}{R}{R}");
+        
+        this.subtype.add("Hydra");
+        this.power = new MageInt(0);
+        this.toughness = new MageInt(0);
+        
+        // Rock Hydra enters the battlefield with X +1/+1 counters on it.
+        this.addAbility(new EntersBattlefieldAbility(new EntersBattlefieldWithXCountersEffect(CounterType.P1P1.createInstance())));
+        // For each 1 damage that would be dealt to Rock Hydra, if it has a +1/+1 counter on it, remove a +1/+1 counter from it and prevent that 1 damage.
+        this.addAbility(new SimpleStaticAbility(Zone.BATTLEFIELD, new RockHydraEffect()));
+        // {R}: Prevent the next 1 damage that would be dealt to Rock Hydra this turn.
+        this.addAbility(new SimpleActivatedAbility(Zone.BATTLEFIELD, new PreventDamageToSourceEffect(Duration.EndOfTurn, 1), new ManaCostsImpl("{R}")));
+        // {R}{R}{R}: Put a +1/+1 counter on Rock Hydra. Activate this ability only during your upkeep.
+        this.addAbility(new ConditionalActivatedAbility(Zone.BATTLEFIELD, new AddCountersSourceEffect(CounterType.P1P1.createInstance(1)), new ManaCostsImpl("{R}{R}{R}"), new IsStepCondition(PhaseStep.UPKEEP), null));
+    }
+
+    public RockHydra(final RockHydra card) {
+        super(card);
+    }
+
+    @Override
+    public RockHydra copy() {
+        return new RockHydra(this);
+    }
+    
+    class RockHydraEffect extends PreventionEffectImpl {
+
+        public RockHydraEffect() {
+            super(Duration.WhileOnBattlefield, Integer.MAX_VALUE, false, false);
+            staticText = "For each 1 damage that would be dealt to Rock Hydra, if it has a +1/+1 counter on it, remove a +1/+1 counter from it and prevent that 1 damage.";
+        }
+
+        public RockHydraEffect(final RockHydraEffect effect) {
+            super(effect);
+        }
+
+        @Override
+        public RockHydraEffect copy() {
+            return new RockHydraEffect(this);
+        }
+
+        @Override
+        public boolean apply(Game game, Ability source) {
+            return true;
+        }
+
+        @Override
+        public boolean replaceEvent(GameEvent event, Ability source, Game game) {
+            PreventionEffectData preventionEffectData = preventDamageAction(event, source, game);
+            if (preventionEffectData.getPreventedDamage() > 0) {

```

Comment:
You'll still want the preventDamageAction to do the actual damage preventing. Looks fine otherwise. Make sure to test though as fine-looking code isn't always working. ;)